### PR TITLE
HPCC-10709 Allow dafilesrv throttle limit to be configured

### PR DESCRIPTION
**Pull request by jakesmith**
Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday @richardkchapman - please review, but don't merge - need to test further.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

**Comment by jakesmith** [comment]
Have added some more configurability in part to make testing easier and tested various throttling options.

@ghalliday - please review.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: common/remote/sockfile.cpp [hunk 1]
```diff
@@ -3430,6 +3414,34 @@ class CRemoteFileServer : public CInterface, implements IRemoteFileServer, imple
...
+        loop {
+            if (throttlesem.wait(throttleDelayMs)) {
+                got = true;
+                break;
+            }
+            unsigned cpu = getLatestCPUUsage();
+            PROGLOG("Throttler: transaction delayed (%d%% cpu)", cpu);
+
+            // NB: getLatestCPUUsage() is based on interval monitoring, typically 60 secs
+            if (getLatestCPUUsage()<throttleCPULimit)
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: Minor: Should this use cpu < ...?

File: dali/dafilesrv/dafilesrv.cpp [hunk 2]
```diff
@@ -354,6 +354,10 @@ int main(int argc,char **argv)
     unsigned short  dafsPort;//DAFILESRV_PORT or SECURE_DAFILESRV_PORT
     querySecuritySettings(&useSSL, &dafsPort, &sslCertFile, NULL);
 
+    unsigned parallelRequestLimit = DEFAULT_PARALLELREQUESTLIMIT;;
```
Comment (ghalliday) [inline dali/dafilesrv/dafilesrv.cpp]: trivial: double ;

File: dali/dafilesrv/dafilesrv.cpp [hunk 3]
```diff
@@ -574,6 +578,34 @@ int main(int argc,char **argv)
         lf->setMaxDetail(TopDetail);
         lf->beginLogging();
     }
+
+    Owned<IPropertyTree> env = getHPCCEnvironment();
+    if (env)
+    {
+        IPropertyTree *daFileSrv = env->queryPropTree("Software/DafilesrvProcess");
```
Comment (ghalliday) [inline dali/dafilesrv/dafilesrv.cpp]: Question: Can there only be a single instance?

**Comment by ghalliday** [comment]
@richardkchapman looks good - some very minor comments.

File: dali/dafilesrv/dafilesrv.cpp (hunk 3 above)
Comment (jakesmith) [inline dali/dafilesrv/dafilesrv.cpp]: I think there always has been in every environment ever created to date, but..
We have talked about having multiple, e.g. on different ports, and the environment xml could had >1 I suppose.

I should use the 'instance name' is passed (which it is in our installation) to ensure I pick up the right one.
I'll change.

File: dali/dafilesrv/dafilesrv.cpp (hunk 2 above)
Comment (jakesmith) [inline dali/dafilesrv/dafilesrv.cpp]: will fix

File: common/remote/sockfile.cpp (hunk 1 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: yes probably.. although it's cached in the call so makes no practical difference - apart from being non-obvious why coded like this.
Will change.

**Comment by jakesmith** [comment]
@ghalliday - pushed with changes

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

**Comment by ghalliday** [comment]
@richardkchapman looks good.  Please sanity check.

File: dali/dfuplus/dfuplus.cpp [hunk 4]
```diff
@@ -64,7 +64,7 @@ class CDafsThread: public Thread
         else
             listenep.getUrlStr(eps);
         enableDafsAuthentication(requireauthenticate);
-        server.setown(createRemoteFileServer());
+        server.setown(createRemoteFileServer()); // no throttle limitting
```
Comment (richardkchapman) [inline dali/dfuplus/dfuplus.cpp]: Actually will use default throttling I think - is that wise? And if so should probably change the comment.

File: dali/dafilesrv/dafilesrv.cpp [hunk 5]
```diff
@@ -548,7 +552,7 @@ int main(int argc,char **argv)
                 PROGLOG("Version: %s", verstring);
                 PROGLOG("Authentication:%s required",requireauthenticate?"":" not");
                 PROGLOG(DAFS_SERVICE_DISPLAY_NAME " Running");
-                server.setown(createRemoteFileServer());
+                server.setown(createRemoteFileServer(parallelRequestLimit, throttleDelayMs, throttleCPULimit));
```
Comment (richardkchapman) [inline dali/dafilesrv/dafilesrv.cpp]: Will these values have been read from anywhere in the isDaemon case?

File: dali/dfuplus/dfuplus.cpp (hunk 4 above)
Comment (jakesmith) [inline dali/dfuplus/dfuplus.cpp]: I meant it to use none here, hence the comment.
I think I change the prototype to use defaults but not this call.

File: dali/dafilesrv/dafilesrv.cpp (hunk 5 above)
Comment (jakesmith) [inline dali/dafilesrv/dafilesrv.cpp]: No, so will be defaults.

**Comment by jakesmith** [comment]
Closing to rebase onto 5.0.16

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

**Pull request closed without merging**